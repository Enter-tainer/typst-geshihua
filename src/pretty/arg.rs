use pretty::{Arena, DocAllocator};
use typst_syntax::{ast::*, SyntaxKind};

use super::{util::is_comment_node, ArenaDoc, PrettyPrinter};

// Handles arg-like expressions, where the comments can only appear in the middle.
pub struct ArgStylist<'a> {
    arena: &'a Arena<'a>,
    printer: &'a PrettyPrinter<'a>,
    doc: ArenaDoc<'a>,
    leading: bool,
}

impl<'a> ArgStylist<'a> {
    pub fn new(printer: &'a PrettyPrinter<'a>) -> Self {
        Self {
            arena: &printer.arena,
            doc: printer.arena.nil(),
            printer,
            leading: true,
        }
    }

    pub fn convert_named(mut self, named: Named<'a>) -> ArenaDoc<'a> {
        // We put a space only before expr and comment when not lead
        for node in named.to_untyped().children() {
            if let Some(expr) = node.cast() {
                self.push_leading_space();
                self.push_doc(self.printer.convert_expr(expr));
            } else if let Some(pattern) = node.cast() {
                self.push_leading_space();
                self.push_doc(self.printer.convert_pattern(pattern));
            } else if node.kind() == SyntaxKind::Colon {
                self.push_doc(self.arena.text(":"));
            } else if node.kind() == SyntaxKind::Hash {
                self.push_leading_space();
                self.push_doc(self.arena.text("#"));
                self.leading = true;
                continue;
            } else if is_comment_node(node) {
                self.push_leading_space();
                self.push_doc(self.printer.convert_comment(node));
                if node.kind() == SyntaxKind::LineComment {
                    self.push_doc(self.arena.hardline());
                    self.leading = true;
                    continue;
                }
            } else {
                // We ignore line break here.
                continue;
            }
            self.leading = false;
        }

        self.doc.group()
    }

    pub fn convert_keyed(mut self, keyed: Keyed<'a>) -> ArenaDoc<'a> {
        for node in keyed.to_untyped().children() {
            if let Some(expr) = node.cast() {
                self.push_leading_space();
                self.push_doc(self.printer.convert_expr(expr));
            } else if node.kind() == SyntaxKind::Colon {
                self.push_doc(self.arena.text(":"));
            } else if is_comment_node(node) {
                self.push_leading_space();
                self.push_doc(self.printer.convert_comment(node));
                if node.kind() == SyntaxKind::LineComment {
                    self.push_doc(self.arena.hardline());
                    self.leading = true;
                    continue;
                }
            } else {
                // We ignore line break here.
                continue;
            }
            self.leading = false;
        }

        self.doc.group()
    }

    pub fn convert_spread(mut self, spread: Spread<'a>) -> ArenaDoc<'a> {
        for node in spread.to_untyped().children() {
            if let Some(expr) = node.cast() {
                self.push_doc(self.printer.convert_expr(expr));
            } else if node.kind() == SyntaxKind::Dots {
                self.push_doc(self.arena.text(".."));
            } else if is_comment_node(node) {
                self.push_doc(self.printer.convert_comment(node));
                if node.kind() == SyntaxKind::LineComment {
                    self.push_leading_space();
                    self.push_doc(self.arena.hardline());
                    self.leading = true;
                    continue;
                }
            } else {
                // We ignore line break here.
                continue;
            }
            self.leading = false;
        }

        self.doc.group()
    }

    fn push_doc(&mut self, doc: ArenaDoc<'a>) {
        self.doc += doc;
    }

    /// Push a space if not leading.
    fn push_leading_space(&mut self) {
        if !self.leading {
            self.doc += self.arena.space();
        }
    }
}
