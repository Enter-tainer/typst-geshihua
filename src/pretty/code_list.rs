use typst_syntax::{ast::*, SyntaxKind};

use super::{
    list::{ListStyle, ListStylist},
    util::is_only_one_and,
    ArenaDoc, PrettyPrinter,
};

impl<'a> PrettyPrinter<'a> {
    pub(super) fn convert_array(&'a self, array: Array<'a>) -> ArenaDoc<'a> {
        ListStylist::new(self)
            .process_list(array.to_untyped(), |node| self.convert_array_item(node))
            .print_doc(ListStyle {
                separator: ",",
                delim: ("(", ")"),
                add_space_if_empty: false,
                add_trailing_sep_single: true,
                omit_delim_single: false,
                omit_delim_flat: false,
            })
    }

    pub(super) fn convert_dict(&'a self, dict: Dict<'a>) -> ArenaDoc<'a> {
        let all_spread = dict.items().all(|item| matches!(item, DictItem::Spread(_)));

        ListStylist::new(self)
            .process_list(dict.to_untyped(), |node| self.convert_dict_item(node))
            .print_doc(ListStyle {
                separator: ",",
                delim: (if all_spread { "(:" } else { "(" }, ")"),
                add_space_if_empty: false,
                add_trailing_sep_single: false,
                omit_delim_single: false,
                omit_delim_flat: false,
            })
    }

    pub(super) fn convert_destructuring(
        &'a self,
        destructuring: Destructuring<'a>,
    ) -> ArenaDoc<'a> {
        let only_one_pattern = is_only_one_and(destructuring.items(), |it| {
            matches!(*it, DestructuringItem::Pattern(_))
        });

        ListStylist::new(self)
            .process_list(destructuring.to_untyped(), |node| {
                self.convert_destructuring_item(node)
            })
            .always_fold_if(|| only_one_pattern)
            .print_doc(ListStyle {
                separator: ",",
                delim: ("(", ")"),
                add_space_if_empty: false,
                add_trailing_sep_single: only_one_pattern,
                omit_delim_single: false,
                omit_delim_flat: false,
            })
    }

    pub(super) fn convert_params(&'a self, params: Params<'a>, is_unnamed: bool) -> ArenaDoc<'a> {
        let is_single_simple = is_unnamed
            && is_only_one_and(params.children(), |it| {
                matches!(
                    *it,
                    Param::Pos(Pattern::Normal(_)) | Param::Pos(Pattern::Placeholder(_))
                )
            });

        ListStylist::new(self)
            .process_list(params.to_untyped(), |node| self.convert_param(node))
            .always_fold_if(|| is_single_simple)
            .print_doc(ListStyle {
                separator: ",",
                delim: ("(", ")"),
                add_space_if_empty: false,
                add_trailing_sep_single: false,
                omit_delim_single: is_single_simple,
                omit_delim_flat: false,
            })
    }

    pub(super) fn convert_import_items(&'a self, import_items: ImportItems<'a>) -> ArenaDoc<'a> {
        // Note that `ImportItem` does not implement `AstNode`.
        ListStylist::new(self)
            .process_list_impl(import_items.to_untyped(), |child| match child.kind() {
                SyntaxKind::RenamedImportItem => child
                    .cast()
                    .map(|item| self.convert_import_item_renamed(item)),
                SyntaxKind::ImportItemPath => {
                    child.cast().map(|item| self.convert_import_item_path(item))
                }
                _ => Option::None,
            })
            .print_doc(ListStyle {
                separator: ",",
                delim: ("(", ")"),
                add_space_if_empty: false,
                add_trailing_sep_single: false,
                omit_delim_single: false,
                omit_delim_flat: true,
            })
    }
}
