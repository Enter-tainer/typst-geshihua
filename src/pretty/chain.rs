use itertools::Itertools;
use pretty::DocAllocator;
use std::iter;
use typst_syntax::SyntaxNode;

use super::{ArenaDoc, PrettyPrinter};

enum ChainItem<'a> {
    Commented { body: ArenaDoc<'a> },
}

pub struct ChainStylist<'a> {
    printer: &'a PrettyPrinter<'a>,
    items: Vec<ChainItem<'a>>,
    chain_len: usize,
    op_space: bool,
}

#[derive(Default)]
pub struct ChainStyle {
    /// Do not break line if the chain consists of only one operator.
    pub no_break_single: bool,
}

impl<'a> ChainStylist<'a> {
    pub fn new(printer: &'a PrettyPrinter<'a>) -> Self {
        Self {
            printer,
            items: Default::default(),
            chain_len: 0,
            op_space: false,
        }
    }

    pub fn space_around_op(mut self) -> Self {
        self.op_space = true;
        self
    }

    pub fn process_resolved(
        self,
        nodes: impl Iterator<Item = &'a SyntaxNode>,
        operand_pred: impl Fn(&'a SyntaxNode) -> bool,
        op_pred: impl Fn(&'a SyntaxNode) -> bool,
        rhs_converter: impl Fn(&'a SyntaxNode) -> Option<ArenaDoc<'a>>,
        fallback_converter: impl Fn(&'a SyntaxNode) -> Option<ArenaDoc<'a>>,
    ) -> Self {
        let mut nodes = nodes.collect_vec();
        nodes.reverse();
        self.process(
            nodes,
            operand_pred,
            op_pred,
            rhs_converter,
            fallback_converter,
        )
    }

    pub fn process(
        mut self,
        nodes: Vec<&'a SyntaxNode>,
        operand_pred: impl Fn(&'a SyntaxNode) -> bool,
        op_pred: impl Fn(&'a SyntaxNode) -> bool,
        rhs_converter: impl Fn(&'a SyntaxNode) -> Option<ArenaDoc<'a>>,
        fallback_converter: impl Fn(&'a SyntaxNode) -> Option<ArenaDoc<'a>>,
    ) -> Self {
        let arena = &self.printer.arena;

        let mut doc = arena.nil();
        for node in nodes {
            if operand_pred(node) {
                self.chain_len += 1;
                let mut seen_op = false;
                for child in node.children() {
                    if op_pred(child) {
                        seen_op = true;
                        self.items.push(ChainItem::Commented { body: doc });
                        doc = if self.op_space {
                            arena.text(child.text().as_str()) + " "
                        } else {
                            arena.text(child.text().as_str())
                        };
                    } else if seen_op {
                        if let Some(rhs) = rhs_converter(child) {
                            doc += rhs;
                        }
                    }
                }
            } else if let Some(fallback) = fallback_converter(node) {
                doc += fallback;
            }
        }
        self.items.push(ChainItem::Commented { body: doc });

        self
    }

    pub fn print_doc(self, sty: ChainStyle) -> ArenaDoc<'a> {
        let arena = &self.printer.arena;

        let mut docs = vec![];
        for item in self.items {
            match item {
                ChainItem::Commented { body } => docs.push(body),
            }
        }

        let op_sep = if self.op_space {
            arena.line()
        } else {
            arena.line_()
        };
        let first_doc = docs.remove(0);
        let follow_docs = arena.intersperse(docs, op_sep.clone());
        if self.chain_len == 1 && sty.no_break_single {
            (first_doc + follow_docs).group()
        } else {
            (first_doc + (op_sep + follow_docs).nest(2)).group()
        }
    }
}

/// Iterates over nodes in a syntax tree in a depth-first manner.
///
/// This function takes a starting node and an `accepter` function,
/// which determines the next node to visit. It returns an iterator that yields
/// nodes in a depth-first order based on the logic defined in the `accepter`.
///
/// # Parameters
///
/// - `node`: The starting node from which to begin the iteration.
/// - `accepter`: If it returns `Some`, the iterator will visit this next node;
///   if it returns `None`, the iteration will stop for that path.
pub fn iterate_deep_nodes<'a>(
    node: &'a SyntaxNode,
    accepter: impl Fn(&SyntaxNode) -> Option<&SyntaxNode> + 'a,
) -> impl Iterator<Item = &'a SyntaxNode> {
    let mut current = Some(node);
    iter::from_fn(move || {
        let ret = current;
        if let Some(ret) = ret {
            current = accepter(ret);
            Some(ret)
        } else {
            None
        }
    })
}
