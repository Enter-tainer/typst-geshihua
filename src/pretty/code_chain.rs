use pretty::DocAllocator;
use typst_syntax::{ast::*, SyntaxKind, SyntaxNode};

use crate::PrettyPrinter;

use super::{
    chain::{iterate_deep_nodes, ChainStyle, ChainStylist},
    util::has_comment_children,
    ArenaDoc,
};

impl<'a> PrettyPrinter<'a> {
    pub(super) fn convert_field_access(&'a self, field_access: FieldAccess<'a>) -> ArenaDoc<'a> {
        if let Some(res) = self.check_unformattable(field_access.to_untyped()) {
            return res;
        }
        if self.current_mode().is_code() {
            return self.convert_dot_chain(field_access.to_untyped());
        }
        self.convert_expr(field_access.target())
            + self.arena.text(".")
            + self.convert_ident(field_access.field())
    }

    pub(super) fn convert_dot_chain(&'a self, node: &'a SyntaxNode) -> ArenaDoc<'a> {
        if resolve_dot_chain(node).any(has_comment_children) {
            return self.format_disabled(node);
        }

        ChainStylist::new(self)
            .process_resolved(
                resolve_dot_chain(node),
                SyntaxKind::FieldAccess,
                |child| child.kind() == SyntaxKind::Dot,
                |child| child.cast().map(|ident| self.convert_ident(ident)),
                |node| {
                    if let Some(func_call) = node.cast::<FuncCall>() {
                        // There is no comment allowed, so we can directly convert args.
                        Some(self.convert_args(func_call.args()))
                    } else {
                        node.cast().map(|expr| self.convert_expr(expr))
                    }
                },
            )
            .print_doc(ChainStyle {
                no_break_single: true,
            })
    }
}

fn resolve_dot_chain(node: &SyntaxNode) -> impl Iterator<Item = &SyntaxNode> {
    iterate_deep_nodes(node, |current| {
        if let Some(field_access) = current.cast::<FieldAccess>() {
            Some(field_access.target().to_untyped())
        } else if let Some(func_call) = current.cast::<FuncCall>() {
            Some(func_call.callee().to_untyped())
        } else {
            None
        }
    })
}
