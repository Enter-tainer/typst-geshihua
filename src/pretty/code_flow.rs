use pretty::DocAllocator;
use typst_syntax::{ast::*, SyntaxKind};

use super::{flow::FlowItem, util::BoolExt, ArenaDoc, PrettyPrinter};

impl<'a> PrettyPrinter<'a> {
    pub(super) fn convert_named(&'a self, named: Named<'a>) -> ArenaDoc<'a> {
        let mut seen_name = false;
        self.convert_flow_like(named.to_untyped(), |child| {
            if child.kind() == SyntaxKind::Colon {
                FlowItem::tight_spaced(self.arena.text(":"))
            } else if child.kind() == SyntaxKind::Hash {
                // name
                FlowItem::spaced_tight(self.arena.text("#"))
            } else if let Some(expr) = child.cast() {
                // expr
                FlowItem::spaced_before(self.convert_expr(expr), seen_name.replace(true))
            } else if let Some(pattern) = child.cast() {
                // pattern
                FlowItem::spaced(self.convert_pattern(pattern))
            } else {
                FlowItem::none()
            }
        })
    }

    pub(super) fn convert_keyed(&'a self, keyed: Keyed<'a>) -> ArenaDoc<'a> {
        let mut seen_key = false;
        self.convert_flow_like(keyed.to_untyped(), |child| {
            if child.kind() == SyntaxKind::Colon {
                FlowItem::tight_spaced(self.arena.text(":"))
            } else if let Some(expr) = child.cast() {
                // key, expr
                FlowItem::spaced_before(self.convert_expr(expr), seen_key.replace(true))
            } else {
                FlowItem::none()
            }
        })
    }

    pub(super) fn convert_spread(&'a self, spread: Spread<'a>) -> ArenaDoc<'a> {
        self.convert_flow_like(spread.to_untyped(), |child| {
            if child.kind() == SyntaxKind::Dots {
                FlowItem::spaced_tight(self.arena.text(".."))
            } else if let Some(expr) = child.cast() {
                // expr, sink_ident, sink_expr
                FlowItem::tight_spaced(self.convert_expr(expr))
            } else {
                FlowItem::none()
            }
        })
    }
}
