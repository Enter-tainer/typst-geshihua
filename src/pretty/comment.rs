use pretty::BoxDoc;
use typst_syntax::{SyntaxKind, SyntaxNode};

enum CommentStyle {
    Plain,
    Bullet,
}

/// Convert either line comment or block comment.
pub fn comment(node: &SyntaxNode) -> BoxDoc<'_, ()> {
    if node.kind() == SyntaxKind::LineComment {
        line_comment(node)
    } else if node.kind() == SyntaxKind::BlockComment {
        block_comment(node)
    } else {
        unreachable!("the node should not be a comment node!")
    }
}

pub fn line_comment(node: &SyntaxNode) -> BoxDoc<'_, ()> {
    BoxDoc::text(node.text().to_string())
}

pub fn block_comment(node: &SyntaxNode) -> BoxDoc<'_, ()> {
    // Calculate the number of leading spaces except the first line.
    let line_num = node.text().lines().count();
    if line_num == 0 {
        return BoxDoc::text(node.text().as_str());
    }
    // Then the comment is multiline.
    let text = node.text().clone();
    let style = get_comment_style(&text);
    match style {
        CommentStyle::Plain => {
            let leading = get_follow_leading(&text).unwrap();
            BoxDoc::column(move |col| BoxDoc::text(align_multiline(&text, leading, line_num, col)))
        }
        CommentStyle::Bullet => {
            BoxDoc::column(move |col| BoxDoc::text(align_multiline_simple(&text, col)))
        }
    }
}

fn get_comment_style(text: &str) -> CommentStyle {
    if text
        .lines()
        .skip(1)
        .all(|line| line.trim_start().starts_with('*'))
    {
        CommentStyle::Bullet // /*
    } else {
        CommentStyle::Plain // otherwise
    }
}

/// Get the minimum number of leading spaces in all lines except the first.
/// Returns None only when the text is a single line.
fn get_follow_leading(text: &str) -> Option<usize> {
    text.lines()
        .skip(1)
        .map(|line| line.chars().position(|c| c != ' ').unwrap_or(usize::MAX))
        .min()
}

/// For general cases. All lines need to be indented simultaneously.
fn align_multiline(text: &str, leading: usize, line_num: usize, col: usize) -> String {
    if col == leading {
        return text.to_string();
    }
    let offset = col as isize - leading as isize;
    let mut result =
        String::with_capacity((text.len() as isize + (line_num as isize - 1) * offset) as usize);
    if col < leading {
        // need to remove line prefix
        let offset = leading - col;
        for (i, line) in text.lines().enumerate() {
            if i == 0 {
                result.push_str(line);
                continue;
            }
            result.push('\n');
            if line.len() > offset {
                result.push_str(&line[offset..]);
            } // otherwise this line is blank
        }
    } else {
        // need to add spaces
        let offset = col - leading;
        for (i, line) in text.lines().enumerate() {
            if i == 0 {
                result.push_str(line);
                continue;
            }
            result.push('\n');
            result.extend(std::iter::repeat_n(' ', offset));
            result.push_str(line);
        }
    }
    result
}

/// For special cases. All lines can be indented independently.
fn align_multiline_simple(text: &str, col: usize) -> String {
    let mut result = String::new();
    for (i, line) in text.lines().enumerate() {
        if i == 0 {
            result.push_str(line);
            continue;
        }
        result.push('\n');
        result.extend(std::iter::repeat_n(' ', col + 1));
        result.push_str(line.trim_start());
    }
    result
}

#[cfg(test)]
mod tests {
    use crate::pretty::comment::{align_multiline, align_multiline_simple, get_follow_leading};

    #[test]
    fn test_align() {
        let text = "/*
      ---
        --
    ---
      */";
        let leading = get_follow_leading(text).unwrap();
        assert_eq!(leading, 4);
        assert_eq!(
            align_multiline(text, leading, 5, 0),
            "/*
  ---
    --
---
  */"
        );
        assert_eq!(
            align_multiline(text, leading, 5, 8),
            "/*
          ---
            --
        ---
          */"
        );
    }

    #[test]
    fn test_align2() {
        let text = "/*
      *
        *
    *
      */";
        let leading = get_follow_leading(text).unwrap();
        assert_eq!(leading, 4);
        assert_eq!(
            align_multiline_simple(text, 0),
            "/*
 *
 *
 *
 */"
        );
        assert_eq!(
            align_multiline_simple(text, 8),
            "/*
         *
         *
         *
         */"
        );
    }
}
