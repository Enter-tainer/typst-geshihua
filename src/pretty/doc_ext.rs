use pretty::{DocAllocator, DocBuilder};

pub trait DocExt<'a, D, A>
where
    A: 'a,
    D: ?Sized + DocAllocator<'a, A>,
{
    fn repeat_n(self, n: usize) -> Self;
}

impl<'a, D, A> DocExt<'a, D, A> for DocBuilder<'a, D, A>
where
    A: 'a,
    D: ?Sized + DocAllocator<'a, A>,
    DocBuilder<'a, D, A>: Clone,
{
    fn repeat_n(self, n: usize) -> Self {
        let mut doc = self.0.nil();
        for _ in 0..n {
            doc = doc.append(self.clone());
        }
        doc
    }
}
