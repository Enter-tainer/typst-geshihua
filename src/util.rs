use pretty::BoxDoc;

/// A style for formatting items
pub enum FoldStyle {
    /// Fold items if them can fit in a single line
    Fit,
    /// Fold items if there is only one item and it fit in a single line, other wise put each item in a line
    Single,
    /// Never fold items
    Never,
}

pub fn pretty_items<'a>(
    items: &[BoxDoc<'a, ()>],
    single_line_separator: BoxDoc<'a, ()>,
    multi_line_separator: BoxDoc<'a, ()>,
    bracket: (BoxDoc<'a, ()>, BoxDoc<'a, ()>),
    bracket_space: bool,
    fold_style: FoldStyle,
) -> BoxDoc<'a, ()> {
    if items.is_empty() {
        return bracket.0.append(if bracket_space {
            BoxDoc::space().append(bracket.1)
        } else {
            bracket.1
        });
    }
    pretty_items_impl(
        items,
        single_line_separator,
        multi_line_separator,
        bracket,
        bracket_space,
        fold_style,
    )
}

fn pretty_items_impl<'a>(
    items: &[BoxDoc<'a, ()>],
    single_line_separator: BoxDoc<'a, ()>,
    multi_line_separator: BoxDoc<'a, ()>,
    bracket: (BoxDoc<'a, ()>, BoxDoc<'a, ()>),
    bracket_space: bool,
    fold_style: FoldStyle,
) -> BoxDoc<'a, ()> {
    let (left, right) = bracket;
    let flat: BoxDoc<'a, ()> = {
        let inner = BoxDoc::intersperse(items.iter().cloned(), single_line_separator);
        let (left, right) = if bracket_space {
            (
                left.clone().append(BoxDoc::space()),
                BoxDoc::space().append(right.clone()),
            )
        } else {
            (left.clone(), right.clone())
        };
        left.append(inner).append(right)
    };
    let multi = {
        let mut inner = BoxDoc::nil();
        for item in items {
            inner = inner
                .append(item.clone())
                .append(multi_line_separator.clone().append(BoxDoc::hardline()));
        }
        let doc = BoxDoc::hardline().append(inner).nest(2);
        left.append(doc).append(right)
    };
    let auto_items = multi.clone().flat_alt(flat).group();
    match fold_style {
        FoldStyle::Fit => auto_items,
        FoldStyle::Single => {
            if items.len() == 1 {
                auto_items
            } else {
                multi
            }
        }
        FoldStyle::Never => multi,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_pretty_items_fit() {
        let strs = ["123", "12345", "1234", "1234567"];
        let docs: Vec<_> = strs.iter().map(|s| BoxDoc::text(s.to_string())).collect();
        let outer = pretty_items(
            &docs,
            BoxDoc::text(",").append(BoxDoc::space()),
            BoxDoc::text(","),
            (BoxDoc::text("["), BoxDoc::text("]")),
            false,
            FoldStyle::Fit,
        );
        insta::assert_debug_snapshot!(outer);
        insta::assert_snapshot!(outer.pretty(10).to_string());
        insta::assert_snapshot!(outer.pretty(80).to_string());
    }

    #[test]
    fn test_pretty_items_single() {
        let strs = ["let a = 1"];
        let docs: Vec<_> = strs.iter().map(|s| BoxDoc::text(s.to_string())).collect();
        let outer = pretty_items(
            &docs,
            BoxDoc::text(";").append(BoxDoc::space()),
            BoxDoc::text(";"),
            (BoxDoc::text("{"), BoxDoc::text("}")),
            true,
            FoldStyle::Single,
        );
        insta::assert_debug_snapshot!(outer);
        insta::assert_snapshot!(outer.pretty(10).to_string());
        insta::assert_snapshot!(outer.pretty(80).to_string());
    }
}
