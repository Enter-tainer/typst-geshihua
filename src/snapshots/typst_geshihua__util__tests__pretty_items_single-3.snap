---
source: src/util.rs
expression: outer.pretty(80).to_string()
---
{ let a = 1 }
